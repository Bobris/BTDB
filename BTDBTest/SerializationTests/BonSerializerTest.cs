using System;
using Assent;
using BTDB;
using BTDB.Bon;
using BTDB.Serialization;
using Xunit;

namespace BTDBTest.SerializationTests;

public class BonSerializerTest
{
    [Generate]
    public class AllSupportedTypes
    {
        public string Str;
        public bool Bool;
        public sbyte Int8;
        public short Int16;
        public int Int32;
        public long Int64;
        public byte UInt8;
        public ushort UInt16;
        public uint UInt32;
        public ulong UInt64;
        public DateTime DateTime;
        public Guid Guid;
        public Half Float16;
        public float Float32;
        public double Float64;
        public AllSupportedTypes? Self;
    }

    [Fact]
    public void SerializeDeserializeAllSupportedTypes()
    {
        var obj = new AllSupportedTypes
        {
            Str = "Hello", Bool = true, Int8 = 42, Int16 = 1234, Int32 = 12345678, Int64 = long.MaxValue,
            UInt8 = 42, UInt16 = 1234, UInt32 = 12345678, UInt64 = ulong.MaxValue, DateTime = new(2024, 2, 11, 14, 4, 30),
            Guid = Guid.Parse("9e251065-0873-49bc-8fd9-266cc9aa39d3"), Float16 = (Half) 3.14, Float32 = 3.14f, Float64 = Math.PI,
            Self = new ()
        };
        var builder = new BonBuilder();
        BonSerializerFactory.Serialize(ref builder, obj);
        var bon = new Bon(builder.FinishAsMemory());
        this.Assent(bon.DumpToJson());
    }
}
