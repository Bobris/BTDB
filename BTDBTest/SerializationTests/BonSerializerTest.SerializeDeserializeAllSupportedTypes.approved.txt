{
  "__type__": "BTDBTest.SerializationTests.AllSupportedTypes",
  "Str": "Hello",
  "Bool": true,
  "Int8": 42,
  "Int16": 1234,
  "Int32": 12345678,
  "Int64": 9223372036854775807,
  "UInt8": 42,
  "UInt16": 1234,
  "UInt32": 12345678,
  "UInt64": 18446744073709551615,
  "DateTime": "2024-02-11T14:04:30.0000000",
  "Guid": "9e251065-0873-49bc-8fd9-266cc9aa39d3",
  "Float16": 3.140625,
  "Float32": 3.140000104904175,
  "Float64": 3.141592653589793,
  "Self": {
    "__type__": "BTDBTest.SerializationTests.AllSupportedTypes",
    "Str": null,
    "Bool": false,
    "Int8": 0,
    "Int16": 0,
    "Int32": 0,
    "Int64": 0,
    "UInt8": 0,
    "UInt16": 0,
    "UInt32": 0,
    "UInt64": 0,
    "DateTime": "0001-01-01T00:00:00.0000000",
    "Guid": "00000000-0000-0000-0000-000000000000",
    "Float16": 0,
    "Float32": 0,
    "Float64": 0,
    "Self": null
  }
}